### PR TITLE
Media: async upload media library grid item update

### LINKED ISSUE #6518: Media library: no updates on connectivity lost
### Expected behavior
When an item is uploading and connectivity is lost, these items should change to "RETRY" automatically.

### Actual behavior
They still show "UPLOADING" - but if you scroll even just a bit to get those items off the screen and back within the viewport again, they show "RETRY" correctly.

### Steps to reproduce the behavior
1. go to the media library
2. start uploading a media item
3. turn airplane mode ON
4. see it shows a toast “connectivity lost” but the item keeps saying “UPLOADING”.
5. If you scroll the list a bit to force update, it works correctly.

### PR DESCRIPTION
**Pull request by mzorz**
Fixes #6518 

The problem is that in `MediaGridAdapter` we check for item equality before actually attempting to update an item in the list. That happens in `updateMediaItem`  through the `equals` method implemented in FluxC for a given `MediaModel`, which checks for 2 items being equal by means of being the very same object or having all the same property values.
When an upload is in progress, FluxC reuses the same media object with each new `onMediaUploaded` event, thus it changes the status but the object reference remains the same, and so it doesn’t end up refreshing the item in the adapter (because comparing an object to itself returns `true` for the `equals()`  implementation).

That specifically is addressed by commit 5b1de02, by adding a flag to force an item update in the grid adapter, so when we have an error event in `onMediaUploaded` or we have a change in a `MediaModel` as reported by `OnMediaChanged`, we force the item to be updated.

I decided to add this flag instead of [removing the `==` equality in FluxC ](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/develop/fluxc/src/main/java/org/wordpress/android/fluxc/model/MediaModel.java#L93)because a) it is true this is only an exception to the general rule that "if an item doesn't change we don't need to update it" and b) the `equals` implementation in FluxC seems right to me, and it is only in this particular context where we need to make sure changes to the _same_  object are reflected on the screen.

In commit f0aa65a we also add a couple of Toasts to signal the user _why_  we can't retry if they tap on retry while they still lack a a connection in place.

To test:

1. go to the media library
2. start uploading a media item
3. turn airplane mode ON
4. see it shows a toast “connectivity lost”
5. also observe with this PR the item says "RETRY" instead of being  stuck at “UPLOADING”.

cc @aforcier

### CODE REVIEW

**Comment by mzorz** [comment]
Just tested: with this same fix we're also fixing this other behavior:

> BUG: no thumbnail. start a draft and upload a media item, exit the editor while it’s still uploading and open the Media library. you’ll see the UPLOADING item there, and once it’s finished it doesn’t show the video thumbnail. It does show the thumbnail if you leave the Media library and come back again.

described in https://github.com/wordpress-mobile/WordPress-Android/issues/6515

File: WordPress/src/main/java/org/wordpress/android/ui/media/MediaBrowserActivity.java [hunk 1]
```diff
@@ -900,7 +901,7 @@ private void addMediaToUploadService(@NonNull MediaModel media) {
         // Start the upload service if it's not started and fill the media queue
         if (!NetworkUtils.isNetworkAvailable(this)) {
             AppLog.v(AppLog.T.MEDIA, "Unable to start UploadService, internet connection required.");
-            return;
+            ToastUtils.showToast(this, R.string.no_network_message, ToastUtils.Duration.SHORT);
```
Comment (aforcier) [inline WordPress/src/main/java/org/wordpress/android/ui/media/MediaBrowserActivity.java]: Was dropping the `return` here intentional?

Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/media/MediaBrowserActivity.java]: thanks for having an eye for this - no this was totally accidentaly, got it back in 1ff3e23

**Comment by aforcier** [comment]
Looks good @mzorz - I'm seeing a related issue though, which is also present in `develop` and perhaps warrants a separate PR:

1. In the media library, start uploading a media item
2. Turn airplane mode on
3. The item now says "RETRY"
4. Keep airplane mode on, and tap retry
5. The item says "QUEUED" and you immediately get a network error toast
6. The item stays "QUEUED", until you scroll the list and come back (like in #6518), then it becomes "RETRY"

**Comment by mzorz** [comment]
good catch @aforcier I noticed something  "wrong" with the QUEUED state on retry yesterday, I think this commit 11a1f90 fixes that.

**Comment by aforcier** [comment]
Works great! :shipit:

**Pull request merged**